File: Programmers/BoHwa/21주차/덧칠하기.py
def solution(n, m, section):
    answer = 0
    arr=[True]*(n+1)
    for s in section:
        arr[s]=False
    for s in section:
        if arr[s]==False:
            answer+=1
            for i in range(s,s+m):
                if i>n: return answer
                arr[i]=True

    return answer


'''
다른 풀이
참고 : https://comdoc.tistory.com/entry/%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%A8%B8%EC%8A%A4-%EB%8D%A7%EC%B9%A0%ED%95%98%EA%B8%B0
'''
def solution(n, m, section):
    answer = painted = 0
    for s in section:
        if s>=painted:
            answer+=1
            painted=s+m
    
    return answer

File: Programmers/Segun/21주차/덧칠하기.cpp
1~n 까지 탐색을 하면서 색칠해야되는 구역이라면 m길이만큼 색을 칠하고
칠한 부분 다음부터로 이어서 탐색한다
주어진 section을 for 문으로 탐색하면서 탐색하는 section이 이전에 m 만큼 칠했던 부분에 속하는지 속하지 않는지에따라
해결하면 더 효율적으로 작성할 수 있다.

#include <string>
#include <vector>
#include<unordered_map>

using namespace std;

int solution(int n, int m, vector<int> section) {
    int answer = 0;
    int result=0;
    unordered_map<int,bool> um;
    for(auto x:section)
        um[x]=true;
    for(int i=1;i<=n;i++){
        if(um[i]==true){
            result++;
            for(int j=0;j<m;j++){
                if(um[i]==true){
                    answer++;
                    if(answer==section.size())
                        return result;
                }
                i++;
            }
            i--;
        }
    }
    return answer;
}

-------------------------------------------------------------------
  
  
#include <string>
#include <vector>

using namespace std;

int solution(int n, int m, vector<int> section) {
    int answer = 1;
    int tmp=section[0];
    for(auto x:section){
        if(x<tmp+m)
            continue;
        tmp=x;
        answer++;
    }
    return answer;
}
